test/e2e/kcp_upgrade.go

```diff
@@ -160,6 +158,61 @@ func KCPUpgradeSpec(ctx context.Context, inputGetter func() KCPUpgradeSpecInput)
 		By("PASSED!")
 	})
 
+	It("Should successfully upgrade Machine deployment and remediate with MachineHealthCheck", func() {
+
+		By("Creating a workload cluster")
+		Expect(input.E2EConfig.Variables).To(HaveKey(clusterctl.KubernetesVersion))
+		Expect(input.E2EConfig.Variables).To(HaveKey(clusterctl.CNIPath))
+
+		var mds []*clusterv1.MachineDeployment
+		cluster, controlPlane, mds = clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy: input.BootstrapClusterProxy,
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder:                filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+				Flavor:                   clusterctl.DefaultFlavor,
+				Namespace:                namespace.Name,
+				ClusterName:              fmt.Sprintf("cluster-%s", util.RandomString(6)),
+				KubernetesVersion:        input.GetPreviousKubernetesVersion(),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(1),
+			},
+			CNIManifestPath:              input.E2EConfig.GetCNIPath(),
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+
+		By("Upgrading MachineDeployment's Kubernetes version to a valid version")
+		framework.UpgradeMachineDeploymentAndWait(context.TODO(), framework.UpgradeMachineDeploymentAndWaitInput{
+			ClusterProxy:                input.BootstrapClusterProxy,
+			Cluster:                     cluster,
+			UpgradeVersion:              input.E2EConfig.GetKubernetesVersion(),
+			WaitForMachinesToBeUpgraded: input.E2EConfig.GetIntervals(specName, "wait-machine-upgrade"),
+			MachineDeployments:          mds,
+		})
+
+		By("Creating a MachineHealthCheck and wait for remediation")
+		framework.ApplyMachineHealthCheckAndWait(context.TODO(), framework.ApplyMachineHealthCheckAndWaitInput{
```

Comment:
Have you considered having the MHC as part of the cluster template and use the Discovery and Wait pattern? 
This approach will ensure full portability because all the objects used during the test can be tuned on a provider bases via the cluster-template file.